### PR TITLE
Remove LXD profile prior to retrying creating LXD instance

### PR DESCRIPTION
**Pull request by yhaliaw**
### Overview

On failure of creating LXD instance, the LXD storage pool is removed.

### Rationale

On deployment of the charm, random failure related to creation of LXD instance are observed.
This is believed to be due to LXD storage pool was not properly initialized.
The above change should allow the retry to initialize the LXD storage pool again.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/lxd.py [hunk 1]
```diff
@@ -351,6 +351,51 @@ def create(
             logger.exception("Failed to create LXD profile")
             raise LxdError(f"Unable to create LXD profile {name}") from err
 
+    def get(self, name: str) -> LxdProfile:
+        """Get an LXD profile.
+
+        Args:
+            name: Name of the LXD profile to create.
```
Comment (jdkandersson) [inline src/lxd.py:358]: To create or to get?

File: src/lxd.py [hunk 2]
```diff
@@ -351,6 +351,51 @@ def create(
...
+
+
+class LxdProfile:
+    """LXD profile."""
+
+    def __init__(
+        self,
+        pylxd_profile: pylxd.models.Profile,
+    ):
+        """Instantiate the LXD storage pool.
```
Comment (jdkandersson) [inline src/lxd.py:377]: Instance of a storage pool?

File: src/runner.py
Comment (jdkandersson) [inline src/runner.py:1]: I didn't spot any tests, would it be worth adding something?

**Review by jdkandersson — DISMISSED** [review]

File: tests/integration/test_charm_no_runner.py [hunk 3]
```diff
@@ -194,3 +196,50 @@ async def test_reconcile_runners(model: Model, app_no_runner: Application) -> No
...
+    The two test is combine to maintain no runners in the application after the
+    test.
+    """
+    # Rename since the app will have a runner.
+    app = app_no_runner
+
+    unit = app.units[0]
+
+    # 1.
+    exit_code, _ = run_in_unit(unit, f"rm -rf {GithubRunnerCharm.ram_pool_path}")
```
Comment (cbartz) [inline tests/integration/test_charm_no_runner.py:227]:
```suggestion
    exit_code, _ = await run_in_unit(unit, f"rm -rf {GithubRunnerCharm.ram_pool_path}")
```

File: src/runner.py [hunk 4]
```diff
@@ -248,7 +248,22 @@ def _create_instance(
...
+        try:
+            instance = self._clients.lxd.instances.create(config=instance_config, wait=True)
+        except LxdError:
+            logger.exception(
+                "Removing resource profile and storage profile due to LXD instance create failure"
+            )
+
+            # LxdError on creating LXD instance could be caused by improper initialization of
+            # storage pool. If other runner LXD instance exists then it cannot be the cause.
+            if self._clients.lxd.instances.all():
```
Comment (cbartz) [inline src/runner.py:260]: When I read the comment I would expect a `not self._clients.lxd.instances.all()` .

**Review by cbartz — APPROVED** [review]
LGTM

**Review by jdkandersson — APPROVED** [review]

**Comment by jdkandersson** [comment]
Looks like some tests failed?

**Comment by cbartz** [comment]
> Looks like some tests failed?

The test failure was unrelated to this PR, the re-run fixed it.

**Pull request merged**